### ISSUE TITLE
bugs, bugs, bugs

### ISSUE THREAD

**Issue by timfel** [Wargus/stratagus#150]
(reported as Bug #1518749 by Karol Krenski - https://bugs.launchpad.net/stratagus/+bug/1518749)

I find regular bugs in the game, that narrowing them seems pointless to
 me. Some of the bugs were already reported by me and by other people. I
 can produce these bugs in various linux conditions with ease. I play
 wargus under various ubuntu versions which I compile the fresh
 stratagus/wargus myself. I could attach logs (wargus -p -i), but since
 the bugs seem so easily reproducable and obvious, I think there's no
 point. I will just enumerate the topics:

- [x] I don't remember the last time I finshed the wargus LAN game. Always
 network out of sync / player sent bad command. Do the developers never
 experience this bug that has been here for years? Do the developers ever
 play wargus multiplayer?

- [ ] ~~In preparation room / multiplayer game: when all players are ready and
 I start the game it happens that some players get the screen blank and
 never make it into the game.~~

- [x] In preparation room: human/orc race setup are never respected.

- [x] In preparation room: start with one peasant only not respected, at
 least sometimes.

- [ ] ~~Lot's of maps are said to be for 2 or 3 people, but the numbers are
 very often misleading. If the maps are cumbersome it might be better to
 have only say 10, but quality maps, working as expected. Otherwise
 players need to check this large collection of maps and hope that they
 will find the one which works as expected. I could produce some maps
 which could be distributed with wargus if the maps are problematic.~~

- [x] Diplomacy: ally human-computer in multiplayer game doesn't work.

- [x] Stratagus MetaServer: what is this thing? Does it work or it's just a
 concept of something to come? Perhaps it's too early to have it in the
 menu?

Look, I really hope stratagus/wargus will work great someday. I love the
 features, the high resolution and that it's opensource. I could try the
 regular warcraft from Blizzard, but I still choose to stay with wargus,
 despite the bugs. My kids and my old friends for LAN parties - we all
 hope that someday we can play undisturbed network wargus games.

**Comment by a-detiste** [comment]
Hi,

Feel generally the same in general & about LAN games particularly ... I did spent a lot of my time to help supporting all these old commeciral games in general, but can't help for this one much.

**Comment by Andrettin** [comment]
a-detiste, these multiplayer bugs have been solved in Wyrmsun (a Stratagus engine game), so it would be a matter of backporting those changes. The difficulty would be identifying precisely which changes from Wyrmsun's fork of Stratagus have to be ported for that to work.

**Comment by a-detiste** [comment]
Well, I've been brain-damaged from doing 13 years of SAS/SQL reporting since I'm 19; now I've been getting enough of my mind back toghetter to work on this `game-data-packager` project in Python http://pkg-games.alioth.debian.org/game-data/ for one year, but my C/C++ is still seriously lacking. Let's say I would do that later.

**Comment by Andrettin** [comment]
Ah, I didn't mean to say for you to do it. I plan to do it myself at some point; now that Stratagus is on GitHub, I want to slowly add Wyrmgus' improvements to it.

**Comment by a-detiste** [comment]
Aah, ok that would be great.

**Comment by mimooh** [comment]
Andrettin, you really fixed these issues I reported? That would be lovely to have these fixes in stratagus! My friends and I, we have sentiment for warcraft so we prefer to stay with stratagus/wargus. But at least I watched some Wyrmsun youtube videos :)

**Comment by Andrettin** [comment]
mimooh, most of them, yes.

I'll go one by one:

"I don't remember the last time I finshed the wargus LAN game. Always network out of sync / player sent bad command. Do the developers never experience this bug that has been here for years? Do the developers ever play wargus multiplayer?"

There was an out-of-sync bug, which I managed to fix. It could be ported easily (IIRC) to Stratagus, but it would be better if someone more knowledgeable about multiplayer code could review it.

"In preparation room / multiplayer game: when all players are ready and I start the game it happens that some players get the screen blank and never make it into the game."

This could either be due to the out-of-sync issue above, or be due to playing a map in multiplayer which has only 1 player. I haven't seen this happen with Wyrmsun, 

"In preparation room: human/orc race setup are never respected."

I had the same problem with Wyrmsun, but fixed it.

"In preparation room: start with one peasant only not respected, at least sometimes."

Same as above; I also experienced this issue with Wyrmsun, but fixed it.

"Lot's of maps are said to be for 2 or 3 people, but the numbers are very often misleading. If the maps are cumbersome it might be better to have only say 10, but quality maps, working as expected. Otherwise players need to check this large collection of maps and hope that they will find the one which works as expected. I could produce some maps which could be distributed with wargus if the maps are problematic."

This is an issue with Wargus, rather than Stratagus. Either each map should be changed to have its true player quantity shown, or the Lua code in Wargus should be changed to override the stated player number with the real one.

"Diplomacy: ally human-computer in multiplayer game doesn't work."

Is allying between human and computer players something introduced with Wargus 2.3.0? I began working on Wyrmsun when Wargus was in 2.2.7, so I haven't kept entirely up-to-date on how Wargus is.

"Stratagus MetaServer: what is this thing? Does it work or it's just a concept of something to come? Perhaps it's too early to have it in the menu?"

That feature is sadly incomplete, to my knowledge :(

**Comment by mimooh** [comment]
"There was an out-of-sync bug, which I managed to fix."
Wow! This was here for years! Good job. Can't wait to have it in stratagus. Please let me know when you're done, I am subscribed to this issue. And please fix this one quickly :) The other issues are minor comparing to this one. The other, but less important issue is to be able to choose the races. The rest of the issues - I can live with them.

In Wargus 2.3.0 there's Diplomacy menu where you can check/uncheck your allies and enemies and also have shared vision among your allies. Works for human players, seems clickable for computer players, but doesn't work.

Thanks for your contribution to this great open source project! And thanks to the whole team of wargus/stratagus!

![screen01](https://cloud.githubusercontent.com/assets/14321180/11326995/24cb24d6-9179-11e5-8e80-3fe0225dc139.png)

**Comment by Andrettin** [comment]
I've uploaded the fixes to the multiplayer race selection and the out-of-sync issues to Stratagus now :)

Do you build Stratagus yourself? If so, let me know if everything is working for you now!

About the diplomacy menu, the thing is that when you click that option you are allying yourself with the AI, but it doesn't mean that the AI is allying itself with you. Basically, with that option you set your stance towards a player.

**Comment by a-detiste** [comment]
My daughter is grounded, how will I be able to test ? :wink:

Nope, I'll just build this & run around the flat.

**Comment by mimooh** [comment]
Wow, that was quick! Just compiling, however:

s1:stratagus$ git show
commit 7ca9428b8da3c8565654c0a955afc04d82100af9
Merge: 7d9e670 cbe17ce
Author: Tim Felgentreff <timfelgentreff@gmail.com>
Date:   Mon Nov 23 14:08:06 2015 +0100

    Merge branch 'master' of github.com:Wargus/stratagus

[ 95%] Building CXX object CMakeFiles/stratagus.dir/src/video/font.cpp.o
/usr/bin/x86_64-linux-gnu-g++   -DDEBUG -DHAVE_GETOPT -DHAVE_STRCASESTR -DHAVE_STRNLEN -DPIXMAPS=\"/usr/share/pixmaps\" -DUSE_BZ2LIB -DUSE_FLUIDSYNTH -DUSE_LINUX -DUSE_MIKMOD -DUSE_MNG -DUSE_OPENGL -DUSE_THEORA -DUSE_VORBIS -DUSE_X11 -DUSE_ZLIB -fsigned-char -g -I/home/mimooh/stratagus/src/include -I/home/mimooh/stratagus/src/guichan/include -I/home/mimooh/stratagus/src/guichan/include/guichan -I/home/mimooh/stratagus/obj-x86_64-linux-gnu-dbg -I/usr/include/lua5.1 -I/usr/include/SDL    -o CMakeFiles/stratagus.dir/src/video/font.cpp.o -c /home/mimooh/stratagus/src/video/font.cpp
/home/mimooh/stratagus/src/video/font.cpp: In member function ‘CGraphic* CFont::GetFontColorGraphic(const CFontColor&) const’:
/home/mimooh/stratagus/src/video/font.cpp:469:94: error: cannot pass objects of non-trivially-copyable type ‘const string {aka const class std::basic_string<char>}’ through ‘...’
    fprintf(stderr, "Could not load font color %s for font %s\n", fontColor.Ident, this->Ident);
                                                                                              ^
/home/mimooh/stratagus/src/video/font.cpp:469:94: error: cannot pass objects of non-trivially-copyable type ‘const string {aka const class std::basic_string<char>}’ through ‘...’
/home/mimooh/stratagus/src/video/font.cpp:469:94: warning: format ‘%s’ expects argument of type ‘char*’, but argument 3 has type ‘const string {aka const std::basic_string<char>}’ [-Wformat=]
/home/mimooh/stratagus/src/video/font.cpp:469:94: warning: format ‘%s’ expects argument of type ‘char*’, but argument 4 has type ‘const string {aka const std::basic_string<char>}’ [-Wformat=]


-- 
Karol Kreński

**Comment by Andrettin** [comment]
Thanks, there was another issue with font.cpp, I've fixed it now:
https://github.com/Wargus/stratagus/commit/4ffc265cc55afe217381ae91bffcc799078d0b41

**Comment by a-detiste** [comment]
stratagus{-dbg} is built correclty, but `make metaserver` fails

even if I hide SQLITE, the build fails somewhere else.

```diff
-if(SQLITE_FOUND)
+if(FALSE)

- libsdl1.2-dev (>> 1.2.2), libx11-dev, libsqlite3-dev,
+ libsdl1.2-dev (>> 1.2.2), libx11-dev,
```

So I guess best is first to bring this to debhelper 9 compatibility level !
Will send a P.R. when I got it building again.

-----------------

libsqlite3-dev

```
[  0%] Linking CXX executable metaserver
/usr/bin/ld: cannot open output file metaserver: Is a directory
collect2: error: ld returned 1 exit status
```
or

```
make[3]: Leaving directory '/home/tchet/git/stratagus'
make -f CMakeFiles/metaserver.dir/build.make CMakeFiles/metaserver.dir/build
make[3]: Entering directory '/home/tchet/git/stratagus'
make[3]: *** No rule to make target 'CMakeFiles/metaserver.dir/build'. Arrêt.
make[3]: Leaving directory '/home/tchet/git/stratagus'
CMakeFiles/Makefile2:107: recipe for target 'CMakeFiles/metaserver.dir/all' failed
make[2]: *** [CMakeFiles/metaserver.dir/all] Error 2
make[2]: Leaving directory '/home/tchet/git/stratagus'
CMakeFiles/Makefile2:119: recipe for target 'CMakeFiles/metaserver.dir/rule' failed
make[1]: *** [CMakeFiles/metaserver.dir/rule] Error 2
make[1]: Leaving directory '/home/tchet/git/stratagus'
Makefile:178: recipe for target 'metaserver' failed
```

debuild -us -uc -b gives an other error

```
==================================
-- Configuring done
CMake Error at CMakeLists.txt:1149 (add_executable):
  Cannot find source file:

    metaserver/cmd.cpp

  Tried extensions .c .C .c++ .cc .cpp .cxx .m .M .mm .h .hh .h++ .hm .hpp
  .hxx .in .txx


CMake Error: CMake can not determine linker language for target: metaserver
CMake Error: Cannot determine link language for target "metaserver".
```

**Comment by mimooh** [comment]
Wargus is broken. Cannot build. I just reported it as wargus issue:
https://github.com/Wargus/wargus/issues/133

Which means I have to wait with testing of the new network fix :(

-- 
Karol Kreński

**Comment by Andrettin** [comment]
a-detiste, the metaserver is (to my knowledge) incomplete. cybermind knew more about it, but unfortunately he stopped working on Stratagus many months ago.

**Comment by a-detiste** [comment]
@Andrettin 

Hey, Wyrmsum is awesome, not like these horrible fan-made graphics I saw 15 years ago in ALE Clone (or this maybe https://en.wikipedia.org/wiki/Stratagus#/media/File:BFM-screenshot01.png )

Will clean the debian/ dir a little bit & give a P.R.

**Comment by Andrettin** [comment]
@a-detiste Thank you! I commissioned the graphics for the game from pixel artists to give it a polished look, glad to know you enjoyed it :)

**Comment by micronizerbattheend** [comment]
Could it be that someone finally fixed the most important and long lived wargus/stratagus bug? I'm so excited about it!
I cannot wait to test. Could someone provide a windows binary for it or basic directions to assemble a version with this fix included? I've compiled stratagus for windows before in attempts to fix this bug but now I don't want to risk it!
Andrettin, if the fix works I promise I will buy your game right away! :)

**Comment by timfel** [comment]
@micronizerbattheend Binaries are under releases: https://github.com/Wargus/stratagus/releases. Also for Wargus: https://github.com/Wargus/wargus/releases

**Comment by micronizerbattheend** [comment]
Thanks!

**Comment by micronizerbattheend** [comment]
I'm currently having a bit of trouble with those binaries, as when I try to run them they ask for some DLLs that are not included in the "dependencies.zip" file. (VCRUNTIME140D.dll, MSVCP140D.dll). I tried installing microsoft visual studio 2015 just to get the dlls but since the install of vs requires a massive amount of hard disk space I wanted to know if there's some workaround. Thanks again.

**Comment by Andrettin** [comment]
@micronizerbattheend Thanks for your nice comments!

About the DLL issues you are having, that's strange. It is the sort of thing that would happen if the the executable being compiled were a debug one, but AFAIK a release executable is set to be compiled.

**Comment by micronizerbattheend** [comment]
I got the file at https://github.com/Wargus/wargus/releases/download/master-builds/Wargus-2.4.exe

The errors pop up either when trying to run wartool during install, or wargus.exe after installation

Meanwhile the stratagus download (not wargus) does not have the same problem. So for now I will just "mix and match" putting the stratagus executable on the wargus folder.

**Comment by timfel** [comment]
Ooops, Wargus was building release binaries. I've just updated the CI script, once it runs through (https://ci.appveyor.com/project/timfel/wargus/build/1.0.31), you'll have a release binary for Wargus. Sorry for the inconvenience

**Comment by mimooh** [comment]
Timfel and Andrettin, thanks a lot for making stratagus/wargus better! I
really, really appreciate your great work!

https://launchpad.net/~stratagus/+archive/ubuntu/ppa says:
"This PPA provides daily builds of Stratagus and some Stratagus Games".
So the new debs are built daily and I don't have to build myself
anymore?! That is great news!

When it comes to the "network out of sync" bug I haven't had an
occassion to fully test it yet. But I'll surely share my comments about
how the fixed stratagus behaves now.

**Comment by micronizerbattheend** [comment]
I also thank Timfel and Andrettin and will test as soon as I get in contact with my group of players. I have high hopes! :)

**Comment by mimooh** [comment]
Unfortunatelly, network out of sync is still there :( Hard to say if the network performance improved at all. I can provide weekly logs, because we usually play it weekly. Today we played around 10 games. The very same linux, the very same fresh deb files, quick LAN, 2 players games and 4 players games - all finished with out of sync. Here is one log of 2 players game:

http://www.inf.sgsp.edu.pl/~mimooh/outOfSync_13.dec.zip

**Comment by Andrettin** [comment]
That's really unfortunate :( So it seems that my fix only solved one desync cause.

I'm able to play Wyrmsun without desyncs, though (and haven't gotten any reports of desyncs since I've fixed that issue). One thing that just sprung to my mind - are you all playing with Warcraft 2 data extracted from a CD-ROM of the same language, or of different languages? Also, are you using the Battle.net version, or the DOS one?

**Comment by mimooh** [comment]
There is a computer lab of identical machines that we use. Exactly the
same data everywhere. We extracted the data from DOS warcraft. First
there are "network out of sync" and then games freezes after "player
sent bad command". Dropping some players allows for continuation of the
game.

I once created a map with lots of units, say 200 against 200. Then I
played a network game and made the units move all over the map. I got
desync infos, but the game never broke which would indicate that it's
not "too much traffic" that causes the problem (normal games are under
100 vs 100 units)

I got the feeling today, that I got no single desync info when I played
against myself on two computers, which means I always played on one
client while there were no clicks on the other. And then switch after
~10 seconds. I developed into TownHall level 3 and dragons and then two
kids came to finish the game and quickly they got desyncs. 

I don't think I can do any serious C programming, but perhaps I could be
of some use in debuging this problem by tracing the conditions inside
the game that lead to desync error. There are some things that interest
me:

a) Can I run a network game against AI? :) It would allow for sending
more network traffic than when I play against myself.

b) What is the purpose of 87654321 in log_of_stratagus_0.log?
0 unit-oil-patch 1 P15 Refs 1: 87654321 7,5 0,0

c) Can I somehow replay the game from log_of_stratagus_0.log? Or logs
are just to compared as text?

**Comment by mimooh** [comment]
If these out of sync issues are hard to fix and there may be  more network issues (performance, security) and since metaserver development is not finished, parhaps raknet could be a solution? It has great opinions on the internet.

**Comment by Andrettin** [comment]
Hey mimooh, thanks for the thorough description of how you got the desyncs, that does help quite a bit. Since you didn't get the issue when playing alone, but in the same game the two players later did, then it is possible that it is the messages causing the error. Did you send any messages when you were playing against yourself?

**Comment by mimooh** [comment]
None of us ever sent any messages

**Comment by a-detiste** [comment]
http://us.blizzard.com/en-us/company/careers/posting.html?id=15000XZ

**Comment by micronizerbattheend** [comment]
From my part testing, I only managed to get two short multiplayer games going, with no desyncs yet.  But it might just be because the games lasted under 10 minutes, so I will try to do longer.

However, I'll mention here (because it relates to the original post on top of this issue report), that race selection doesn't work, and neither does diplomacy (in our games players couldn't even share vision).

**Comment by Andrettin** [comment]
I tested diplomacy and settings races with Wyrmsun and both worked every time I tested. It seems quite possible that the issues you're experiencing are due to how Wargus is configured, rather than with Stratagus itself.

Regarding diplomacy - are you sure both of you are setting your stances to the same? Shared vision only works if both players have marked shared vision with each other.

**Comment by mimooh** [comment]
> I tested diplomacy and settings races with Wyrmsun and both worked
> every time I tested. It seems quite possible that the issues you're
> experiencing are due to how Wargus is configured, rather than with
> Stratagus itself.
Race selection in Wargus is not respected, I confirm. We all like orcs
here so I just crated orcs only map and we all play orcs and this is not
much issue to me. But these out of syncs make my heart cry... 

-- 
Karol Kreński

**Comment by micronizerbattheend** [comment]
>We all like orcs here so I just crated orcs only map and we all play orcs

I think it wouldn't be too hard to put a script into the map file that went like "if player1name = "mimooh" then switch all player1peasants with peons" as a temporary workaround for the race selection, but I agree with you that the most important issue here are the desyncs.

**Comment by Kintobor** [comment]
Cyber and I did heaps of testing. I thought we eliminated these desyncs in 2.3. We played FtM on Windows machines with a ping of 400. Can someone confirm that FtM desyncs?

**Comment by DinkyDyeAussie** [comment]
I thought the de-syncs were sorted in 2.3 as well. Everything was working great. Had alot of the old school features almost fully implemented again and now it seems the wargus game has taken a few steps backward.

**Comment by timfel** [comment]
It is weird that race selection is not respected, the code is there to map units from one race to another (in lua) and it is working in my tests. What map are you using, what version of wargus and stratagus (including git rev)?

**Comment by timfel** [comment]
@micronizerbattheend the race selection already is entirely implemented in wargus - see wc2.lua. It uses a mapping between human and orc units and overrides the CreateUnit function to replace units woth their equivalent depending on the race selection.

**Comment by mimooh** [comment]
Could it be that desyncs are linux related then? Hmm... I remember playing LAN games on 2 windows machines sometime early in 2015 (even before v2.3 I think). Even though we clicked the fastest we could, produced hundreds of units, we managed to finish most of the games. But whenever 2 kids play under linux the games break under 30 minutes I'd say. Kids play slow and produce few units and I think out of syncs happen more in quicker/larger games. 

I created a massive map and then played on ubuntu. Two instances of wargus on localhost playing against each other. Within 5 minutes I've got planty out of syncs. After some time it resulted in "player sent bad command" and the player was dropped.

timfel, I will see about race selection withing an hour

**Comment by mimooh** [comment]
I confirm the bug with race selection. Four players map ./maps/skirmish/(4)central-park.smp.gz. 4 clients chose orcs. In game there were 2 orcs and 2 humans. All under ubuntu. I don't know precise github versions, but it's after the commits race related.  It's fresh enough since -G option is there:

student@s13:~$ wargus --version
Stratagus v2.4.0, Copyright (c) 1998-2015 by The Stratagus Project
	-G "options"	Game options (passed to game scripts)

BTW, timfel, the -G option is the most awesome option of them all! :) All multiplayer fans should check it: it allows for setting all game parameters directly from command line, so you don't need to select anything in game praparation room.

**Comment by timfel** [comment]
@mimooh regarding testing out-of-sync bugs, I just pushed a 'dedicated' option that you can pass to the multiplayer server commandline. This way, the server will run player 0 as an AI player. If you launch a game with `wargus -c multiplayer -G server,dedicated,map=islands.smp.gz,numplayers=1` and choose a map that has 8 `person` slots, then the server will run 7 AI players on it, and you can connect as a human player to test the network.

**Comment by mimooh** [comment]
@Tim, as always, I love your -G options :) In fact I once mentioned the
need for having this AI bots for network tests to cheaply create traffic
for me. But it turned out I could do without bots: a massive map would
allow for quick out-of-syncing the game just playing myself in two
windows.

But still this option is useful as it allows for handicapped games. When
I play against my kid he can ally with the AI to beat daddy! :)

-- 
Karol Kreński

**Comment by timfel** [comment]
Ok, there are at least three issues that I can see that prevent race selection from working that all happen to be only in multiplayer games. I fixed two of those locally, I have to go now, but might have a fix for the third (and hopefully last) issue later today

**Comment by timfel** [comment]
Ok, the race selection should work now. The first issue was that the order for the selection dropdown was reversed from the single player variant, but the lua code for swapping the units relies on those indices. The second issue was that the C++ code then subtracted 1 from this index, so rather than just getting the reverse races, you would get default when you select orc, and human when you select human. Finally, the reversal script needs to run on the clients, too, otherwise the different clients see inconsistent initial units, but the GameSettings were not being synchronized correctly, so some C++ code also needed fixing.

**Comment by Andrettin** [comment]
Tim, the 1 has to be decreased from the index, otherwise the race won't be set correctly. The two races are 0 and 1, but Wargus sets them to 1 and 2.

When you fixed the reversal in Wargus, that was AFAIK all that needed to be changed, as in Stratagus itself the race selection worked fine (as can be attested since it worked for other Stratagus games).

**Comment by mizukami999** [comment]
timfel, thank you for 2.4, It's my only hope to get the game running in multiplayer. However, my Wargus crashes when Orc campaig mission VIII starts. The one about the Runestone. I wonder if anyone else has that issue.
BTW can anybody confirm the "out of sync" error no longer crashes the multiplayer session?

**Comment by mimooh** [comment]
Unfortunatelly, "out of sync" is not fixed :(

-- 
Karol Kreński

**Comment by mizukami999** [comment]
And the session crashes? )

**Comment by mimooh** [comment]
It crashes :(

**Comment by mimooh** [comment]
Perhaps "out of sync" is linux specific only. Could someone check it under windows? Just "out of sync" is not what crashes the game. More and more desyncs seem to result in "player sent bad command", which is fatal and then at least some players must be dropped and the game can continue as I can recall.

**Comment by mizukami999** [comment]
Only use it on Windows. I'm no programmer, but I wonder if there's no proper procedure to cure that. I don't know - pause the game and resync everything.

I'll try to test what happens if I lower max unit count in wargus/scripts/stratagus.lua

**Comment by mimooh** [comment]
> Only use it on Windows. I'm no programmer, but I wonder if there's no proper procedure to cure that. I don't know - pause the game and resync everything.
We tried that. Once there were massive desyncs we would stop touching
the mouse, but it wouldn't prevent the crash.

> I'll try to test what happens if I lower max unit count in wargus/scripts/stratagus.lua
Yes, that should lower the desyncing. I have done the opposite: created
a massive map and managed to crash the game quickly (5 minutes perhaps).

-- 
Karol Kreński

**Comment by mizukami999** [comment]
Manually is one thing. What I meant was whether it's hard to fix the network code so that wherever the session starts desyncing, the game just pauses itself and resyncs. Like, reloads the units completely and stuff.

**Comment by mimooh** [comment]
> Manually is one thing. What I meant was whether it's hard to fix the
> network code so that wherever the session starts desyncing, the game
> just pauses itself and resyncs. Like, reloads the units completely and
> stuff.
Wow, interesting! :) I don't speak C, so perhaps someone knowledgable
could comment on that. From a player's perspective I would love that
resolution - minor alternation in my units features shouldn't matter
much.

-- 
Karol Kreński

**Comment by mizukami999** [comment]
Yeah! 
I mean, how great it would be, like, to meet a C speaker, who's, like, "fix a sync problem? - pff. I can fix a thousand sync problems, like, in a minute!"

**Comment by timfel** [comment]
The problem is that none of the people who wrote that network code are still actively working on Stratagus. Understanding it completely is not much easier than re-writing it from scratch, especially since the original design was heavily driven by the desire to support even 28.8 kbit/s modems with minimal lag. Today, this really isn't a requirement anymore, so many things could be simplified. It's just a question of time. And then the other problem is that I cannot even debug this, since I cannot reproduce it :(

**Comment by Andrettin** [comment]
> The problem is that none of the people who wrote that network code are still actively working on Stratagus. Understanding it completely is not much easier than re-writing it from scratch, especially since the original design was heavily driven by the desire to support even 28.8 kbit/s modems with minimal lag. Today, this really isn't a requirement anymore, so many things could be simplified. It's just a question of time. And then the other problem is that I cannot even debug this, since I cannot reproduce it :(

Yeah, specially not being able to reproduce it is a very big hurdle. I cannot reproduce it either, and neither could Kyran and cybermind. The bug seems to affect some people nearly every time they play, and others not at all. Additionally, it could be a Lua issue with Wargus, rather than having to do with Stratagus' C++ code (since for Wyrmsun no one has reported the "player sent bad command" issue since I fixed the problem with the chat messages).

One thing that could help though is making the "DebugPrint" calls which indicate errors print the same text to stderr.txt as well, so that when mimooh and others play and get the issue, it is possible that there will be some indication of what the error was in stderr.txt.

**Comment by mimooh** [comment]
> not much easier than re-writing it from scratch, especially since the
> original design was heavily driven by the desire to support even 28.8
> kbit/s modems with minimal lag. Today, this really isn't a requirement
> anymore, so many things could be simplified. 
If rewriting would be considered I once again mention this raknet
thing.

> It's just a question of time. 
That sounds optimistic :)

> And then the other problem is that I cannot even debug this, since I
> cannot reproduce it :(
If you tried and haven't got affected then maybe it is OS related?
Perhaps it's not that much broken - developers say they have thoroughly
tested the network gameplay and it worked for them. I tried under
various ubuntu versions over last 3 years or so - it crashes. Could it
be the problem of the version of warcraft2 dataset? I can privately
share my warcraft2 copy. 

-- 
Karol Kreński

**Comment by mimooh** [comment]
> Yeah, specially not being able to reproduce it is a very big hurdle. I
> cannot reproduce it either, and neither could Kyran and cybermind. The
> bug seems to affect some people nearly every time they play, and
> others not at all. Additionally, it could be a Lua issue with Wargus,
Wow, I sort of thought that the desyncs affect the majority of users and
they are just hard to fix. Fist thing is I should not discourage new
users by saying "out of sync" is always there - you should have told me
that! :) The other thing is that I can help with narrowing the
conditions for the game to desync. Specially thru this "DebugPrint"
mechnism that you mention. You think trying various warcraft2 datasets
would make sense too? 

-- 
Karol Kreński

**Comment by timfel** [comment]
This didn't leave me any peace, and now I was able to reproduce it with Warcraft 1 dataset (running a network game with ~200 units, 12 players, and 256x256 map between mixed Ubuntu & Windows computers). Anyway, I pushed a small change to force a Network recovery when this happens, and then only print the error message if it happens again, and otherwise report if the network is back in sync on the next message. And indeed, I now get alternating "Recovering" ... "Back in sync" messages. This might allow the game to continue, but the problem is definitely deeper - almost every second command triggers an out-of-sync, pretty sure that's not the way it's intended to be.

**Comment by timfel** [comment]
@Andrettin can you take a look at https://github.com/Wargus/stratagus/commit/7ee413e51faa6a785386b0b7112d45d8174276cc and tell me if that may be reasonable? Setting the `NetworkInSync` to false triggers a call to `NetworkRecover` on the next game loop.

**Comment by mimooh** [comment]
> This didn't leave me any peace, and now I was able to reproduce it
> with Warcraft 1 dataset (running a network game with ~200 units, 12
> players, and 256x256 map between mixed Ubuntu & Windows computers).
Hah, now you see yourself! Tim, you know you will be the superhero if
you fix this! :)

-- 
Karol Kreński

**Comment by mimooh** [comment]
> And indeed, I now get alternating "Recovering" ... "Back in sync" messages.
> This might allow the game to continue, but the problem is definitely deeper -
> almost every second command triggers an out-of-sync, pretty sure that's not the
> way it's intended to be.
You mean: "how come that right after 'Back in sync' the game is immediately out of sync on the next game loop again", right? This would indicate that NetworkRecover doesn't really recover at all. Can't wait to find out what you found there :)

**Comment by timfel** [comment]
Ok, after a few more experiments, I can now reliably trigger this within the first 200 cycles even on a localhost connection. Both the syncseed and the synchash diverge. The seed seems to not do much other than test that all clients have the same overflow behavior, but it looks like it actually relies on undefined behavior and I believe it will have problems if you use different word sizes or compilers on different platforms. This part is easily fixed. The sync hash encodes actions taken over time, and right now it diverges even if no human player is present, which might indicate the AI has some non-determinism. Indeed, I was able to trigger some random behavior in worker selection for harvesting, which (from what I can tell) causes the sync hash to diverge immediately. This might not be the only thing. So, to fix this properly, a more thorough code review might be in order.

**Comment by mimooh** [comment]
Tim, it's a huge step forward. Thanks for you contribution!

**Comment by mizukami999** [comment]
>  I was able to trigger some random behavior in worker selection for harvesting, which (from what I can tell) causes the sync hash to diverge immediately.

I came up with an experiment idea:
start a game with 2 players, Reveal map: on, fog: off. 
populate map with peasants, make them harvest lots of stuff. 
Record video of player a's peasants harvesting with fraps a) on the local machine b) on a remote machine 

then compare the videos to see if at a certain point peasant positions are no longer the same, and how early.

IF the problem is with AI behaviour, then it can be recreated without too many units on the map - only a small but crowded part of the map is needed

**Comment by timfel** [comment]
That's the kind of map I'm using right now - except that all the peasants are AI controlled, and the two human players don't do anything. It takes about 10 seconds to visibly see different peasants moving towards harvesting☺

**Comment by timfel** [comment]
And since (afaict) the syncHash relies on the idea that the AI deterministically executes the same on all machines, this is would definitely be a problem.

**Comment by mizukami999** [comment]
FACEPALM!
I guess the normal procedure would be to have the server decide how the units move, and just inform all other machines on what really happens. Only the server is supposed to decide what actally happens. Only the server moves the units...

Clients may simulate unit movement for smooth performance, but then the update cometh, and the client corrects itself the peasant teleports to where he's supposed to be...

This reminds me of how I once had a nice Operation Flashpoint "game of friendship", where I saved my truck convoy on my machine, and my pal destroyed it on his. We both had screenshots as proof.

**Comment by timfel** [comment]
@mizukami999 well, it's not necessarily a bad idea, the nice thing about this architecture is that the server can crash or leave and we don't have to have code to re-negotiate who becomes the new server.

**Comment by mizukami999** [comment]
It's true for advanced projects. But for something that keeps crashing because it can't synchronize units since multiplayer was introduced back in 1812... what people demand is order, clarity and purely totalitarian approach)

**Comment by mizukami999** [comment]
Allowing multiplayer savegames could be an even better fix for server crashes...

**Comment by mimooh** [comment]
In case it matters: games between humans only (no AI) also desync.

-- 
Karol Kreński

**Comment by mizukami999** [comment]
That's because AI is also about pathfinding: peasants colliding into one another, finding themselves in parallel universes.

**Comment by timfel** [comment]
It may also be because of the random seed. It would be useful to get a log for such a game with no AI that desyncs (start all clients with -p)

**Comment by mimooh** [comment]
> I guess the normal procedure would be to have the server decide how
> the units move, and just inform all other machines on what really
> happens. Only the server is supposed to decide what actally happens.
> Only the server moves the units...
The design is described in the beginning of network.cpp. If this is up
to date then I'd say there's no central server in-game - the clients
communicate directly with each other. I could do with network traffic
workflow debugging if that would be required - if we want to know for
sure if there's p2p.

** @subsection sc_vs_p2p server/client vs. peer to peer
**
** @li server to client
**
** The player input is send to the server. The server collects the input
** of all players and than send the commands to all clients.
**
** @li peer to peer (p2p)
**
** The player input is direct send to all others clients in game.
**
** p2p has the advantage of a smaller lag, but needs a higher bandwidth
** by the clients.
**
** p2p has been chosen for in-game.
** s/c for the preparing room.

@Tim, concerning the hashes. So is it that each client has it's game
state and then runs like md5() (some other function here) on this game
state which results in a hash of sort A01BFF2138 that I can see in log
files. And then these hashes are compared amongst each other and this is
how we know the game states differ? Then we say the game desyncs, right?

-- 
Karol Kreński

**Comment by mimooh** [comment]
> It would be useful to get a log for such a game with no AI that
> desyncs (start all clients with -p)
Higher in this thread (13 Dec 2015) you have exactly such a log of mine:
http://www.inf.sgsp.edu.pl/~mimooh/outOfSync_13.dec.zip

-- 
Karol Kreński

**Comment by timfel** [comment]
@mimooh That log does not contain any desync messages. These messages look like so:
```
Network out of sync fabbd813!=fabcd712! 1278312038!=3820182023! Cycle 424
```

**Comment by timfel** [comment]
As for the hashes, the has is basically what you describe, but we also generate a desync message if the RandomSeed is out of sync - these are different. The random seed is a pseudo-random generated number that starts of in the same state for all machines and is updated each time it is used (such as to determine unit directions, movement of spells like whirlwind, ...). If the hash goes out of sync, it indicates that the clients are executing different unit actions or have different units, if the random seed goes out of sync, it indicates that the clients use different randomness (for example, one uses randomness in their lua scripts while the other doesn't). Both are a problem. But the way it is, random seed can already go out of sync simply because of undefined behavior and different word sizes, so that is a bug in the engine.

**Comment by timfel** [comment]
I am working on a branch (tim/network-issues) to check these issues out. If you feel adventurous, you can try to compile and use it yourself. But I'm just poking at various parts of the code right now to see what happens, so no promises.

**Comment by mimooh** [comment]
In the evening I could create such logs. Strange, I thought I used the
-p switch...

-- 
Karol Kreński

**Comment by mimooh** [comment]
> like whirlwind, ...). If the hash goes out of sync, it indicates that
> the clients are executing different unit actions or have different
> units, if the random seed goes out of sync, it indicates that the
> clients use different randomness (for example, one uses randomness in
I admire how you managed to figure out the design from just reading the
code. Out of curosity and based on these hints I will have a look (just
read) at the code. Thanks Tim.

-- 
Karol Kreński

**Comment by Andrettin** [comment]
> The random seed is a pseudo-random generated number that starts off in the same state for all machines and is updated each time it is used (such as to determine unit directions, movement of spells like whirlwind, ...).

I wonder - that could be the reason the issue hasn't happened with Wyrmsun, since it doesn't have a whirlwind-like spell.

**Comment by timfel** [comment]
@Andrettin the issue happens in my test map that *only* has workers within the first 300 cycles. No spells are being cast.

**Comment by Andrettin** [comment]
Ah yes :(

**Comment by mizukami999** [comment]
whirlwind lol
we need to make this spell generate a stop error screen on enemy machine

**Comment by mimooh** [comment]
> whirlwind lol
> one needs to make this spell generate a stop error screen on enemy machine
Hahaha! Indeed :)

-- 
Karol Kreński

**Comment by mimooh** [comment]
network.cpp:
** @li peer to peer (p2p) 
** The player input is direct send to all others clients in game.
** p2p has been chosen for in-game. 

Whatever they mean by the above, based on the network traffic I'd say the clients never comunicate with each other. I ran a 3 players game (server + 2 clients) and all the traffic goes through the server.

(...) But the way it is, random seed can already go out of sync simply because of undefined behavior and different word sizes, so that is a bug in the engine.
@Tim,  surely that's reasonable, but I am probably affected by hash out of sync, since I test on 16 identical machines (hardware and software).

**Comment by timfel** [comment]
@mimooh I tested via localhost connection on the same machine running the exact same binary and randomseed went out of sync... ;)

**Comment by mimooh** [comment]
Aaah, I see. You think your latest fixes under tim/network-issues may be good enough already? You think I should try and break the session as usual? :) Or it's too early for testing?

**Comment by timfel** [comment]
I have not tested the last code change I pushed, and have no time to work on it right now

**Comment by mimooh** [comment]
Ahhh, now I get all this... looks like stratagus follows the ideas described on these two, very interesting sites:
1. http://gafferongames.com/networking-for-game-programmers/what-every-programmer-needs-to-know-about-game-networking/
2. http://www.gamasutra.com/view/feature/131503/1500_archers_on_a_288_network_.php?page=1

I was more familiar with how quake networking works, but RTS games are
different, because game states are too huge to be sent over the network, so:

a) all clients start from very same state
b) all clients send their actions to the network
c) all clients get other clients actions and then they just repeat these actions
locally and all clients should end up in the very same game state.

It is a p2p in a sense, because there's no central place where the global game
state is stored. 

Unfortunatelly, out-of-sync is very challenging:

"At first take it might seem that getting two pieces of identical code to run
the same should be fairly easy and straightforward -- not so. The Microsoft
product manager, Tim Znamenacek, told Mark early on, "In every project, there
is one stubborn bug that goes all the way to the wire -- I think out-of-sync is
going to be it." He was right. The difficulty with finding out-of-sync errors
is that very subtle differences would multiply over time. A deer slightly out
of alignment when the random map was created would forage slightly differently
-- and minutes later a villager would path a tiny bit off, or miss with his
spear and take home no meat. So what showed up as a checksum difference as
different food amounts had a cause that was sometimes puzzling to trace back to
the original cause.

As much as we check-summed the world, the objects, the pathfinding, targeting
and every other system -- it seemed that there was always one more thing that
slipped just under the radar."

and

"Nobody on the Age of Empires development team would argue the need for the
best sync tools possible. As with any project, when you look back on the
development process during a postmortem, some areas always stand out as the
ones you spent the most time on, but could have spent much less time on given
more up-front work. Synchronization debugging was probably at the top of this
list as we started development on RTS3."

**Comment by mimooh** [comment]
"The random seed is a pseudo-random generated number that starts off in the same state for all machines and is updated each time it is used (such as to determine unit directions, movement of spells like whirlwind, ...)"

If we removed the randomness, just for a while if we made it a constant or always increment by 1 (if it cannot be constant), wouldn't it's only effect be the loss of some little randomness in these unit directions and whirlwind etc.?

**Comment by Andrettin** [comment]
> If we removed the randomness, just for a while if we made it a constant or always increment by 1 (if it cannot be constant), wouldn't it's only effect be the loss of some little randomness in these unit directions and whirlwind etc.? 

The randomness is vital - it determines the quantity of damage dealt, for instance. In any case, even if it were removed, that wouldn't really fix the problem - the event that triggered the discrepancy in the random seed would still occur - we just would have more trouble identifying it.

**Comment by mizukami999** [comment]
If I have authority to move my units then i must inform everyone else on where my units are. And they must be updated on remote machines. Otherwise I only send my commands, like "i wish this peasant could go there". Someone who knows better responds, like: sorry to infowm you, but your peasant has been dead for a minute already. Either I execute randomness and inform others on results, or the server does it. I don't think there is any democracy in this procedure. You can simulate whatever you want on your machine for smooth performance, but either you tell the others what really happens with your units, or the server tells you that.

**Comment by mimooh** [comment]
@Andrettin, I went thru lot's of reading about RTS networking since yesterday.
I hope to finally see the big picture, I learnt about the main issues. 

What is the least clear to me is: what the hell is the purpose of randomness in
the first place? Seems like it just enhancement in the game so that you don't
get bored - randomness prevents you from winning against the AI in the very
same clicks over and over, because the damage you make to other units, because
the direction of whirlwinds and perhaps other details are random. Or is
randomness serving some other purposes?

My argument is: Perhaps it makes sense in human vs AI. But in human vs human
multiplayer game there's enough randomness from just the fact that we probably
won't have exactly the same game again - that would be true even for chess -
you and your opponent's moves would produce a different game 99% of the time.

So we could just try to simplify the game and remove randomness and see how
players like it and do they see the change at all in human vs human games. 

If, however we need the randomness, then we invite the low level scarry
monsters. All out-of-sync hell seems to result from the need to produce exactly
the same randomness on each participant computer. 

I see this code in stratagus:

	float radians = deg2rad(MyRand() % 360);

floats... And then I read the comments of IT experts at 
http://gafferongames.com/networking-for-game-programmers/floating-point-determinism/

"People even report different results on the same machine from run to run, and
between debug and release builds. Other folks say that AMDs give different
results to Intel machines, and that SSE results are different from x87."

"Many programmers may not realize that even a program that uses only the numeric
formats and operations prescribed by the IEEE standard can compute different
results on different systems."

"So if you’re trying to floating point determinism; stay away from the intrinsic
hardware calls like (_mm_rsqrt_ss), they’re not the same between Intel and AMD."

"It’s probably time to stop depending on floats to be deterministic on PCs,
though. No new features on 64 bit CPUs are deterministic across models /
manufacturers. All hail the rise of cross platform deterministic int based game
and physics engines!"

"We dealt with this back in 1996 (...) After several attempts to get the
then-current version of MSVC to behave according to IEEE 754 and match the PPC
version, we moved to fixed-point."

As you can see some people propose switching from float to fixed point. Which
only opens another bag of issues, e.g. you need to implement sqrt(), sin() and
resolve divide by 0 and such.

What I liked the most and found only in one place is an interesting
proposition: just hardcode some values in a list and then see about taking from
that list. Because, hey, how many different angles do you need for the
whirlwind direction?

There is the second category of out-of-sync:

"Desyncs are usually programmer error (...) of the most common cases is an
uninitialized variable."

and then "A Demigod Tale" case study at
http://www.gamasutra.com/view/news/126022/Opinion_Synchronous_RTS_Engines_And_A_Tale_of_Desyncs.php

**Comment by a-detiste** [comment]
this thread is awesome

randomness is part of the fun, it's about taking a bit of risk etc...

the only thing I like even more than Warcraft II is "Advance Wars 2" and that would be boring
without some random hit-points

http://jmtd.net/log/deterministic_doom/

for the radian problem: maybe rounding those to closest integer would be ok ?

**Comment by Andrettin** [comment]
> I see this code in stratagus:
> float radians = deg2rad(MyRand() % 360);

The use of "MyRand()" here really drew my attention - this code isn't using SyncRand(), which is the proper function for synchronized randomness. I did a search for MyRand() and found that it is used for unit directions instead of SyncRand()... This will result in units in different computers having different directions. This could very likely be the source of at least one of the desyncs.

**Comment by Andrettin** [comment]
> If, however we need the randomness, then we invite the low level scarry
monsters. All out-of-sync hell seems to result from the need to produce exactly
the same randomness on each participant computer.

I don't think the normal way of using randomness (SyncRand()) is causing any issues. In Wyrmsun entire maps are generated randomly with synchronized randomness, and all players end up with exactly the same map. The thing is, the random seed is changed whenever a random number is generated. so the randomness is probably just telling us that an action happened in one computer but not the other, rather than causing the desync itself.

**Comment by mimooh** [comment]
> Wyrmsun entire maps are generated randomly with synchronized randomness, and
> all players end up with exactly the same map. 

@Andrettin, maybe they just seem the same? Do computers compare them? 

Are mines locations compared across the clients in case of randomly generated
maps? In case they are not, maybe my peasants in my game walk 10 cm and in your
11 cm? At the begining there's no problem for any of us, so the game continues.
But I collect money faster. Then I send you the command that I am building the
Ogre Moud, which you cannot redo in your game, since there's not enough money
in my Town Hall on your computer. 

Heh, I didn't quite get the feeling of theirs "butterfly and tsunami" example
and how hard it is to find the origins of the problem, until I came up with my
own example above :) Checksuming everything we can and comparing seems indeed
essential, because we can probably invent more of the above devil examples.

**Comment by Andrettin** [comment]
> Are mines locations compared across the clients in case of randomly generated
maps? In case they are not, maybe my peasants in my game walk 10 cm and in your
11 cm? At the begining there's no problem for any of us, so the game continues.
But I collect money faster. Then I send you the command that I am building the
Ogre Moud, which you cannot redo in your game, since there's not enough money
in my Town Hall on your computer.

If the random seed were different, then the building would likely be in a completely different location, rather than just slightly different. Also, since the engine is tile-based a position difference would have quite drastic effects on gameplay, which would quickly cause a desync.

In any case, I've pushed a commit fixing the MyRand() issue; this removes at least one desync cause.

**Comment by mimooh** [comment]
> If the random seed were different, then the building would likely be
> in a completely different location, rather than just slightly
> different. 
I agree. I am not sure I am talking about randomSeed, because it is an
integer. I am just saying that the places where floats appear should be
double checked since they are not guarantied to produce the same
calculations on different computers. Enough that one client's AI makes
float z=sin(x) and sends a critter (I would just remove all these
annoying bastards out of the game! :) at 0.0001 radian too much to the
north and after 10 minutes someone cannot build a farm there. And who
controlls critters? AI controlls them. So they wouldn't be sent through
network, right? So my game has my critters, and your game has yours and
we never check against their locations. The next test I try will be the
map full of critters :)

And, but this is just a guess, the floats could work better for some
operations, maybe for sqrt() - maybe there the difference in floats
doesn't matter *that much*. And maybe sqrt() is for map generation, so
maps are generated just fine. But maybe sin() has such an implementation
on my hardware/software, that my results are just too much different
from yours.

> In any case, I've pushed a commit fixing the MyRand() issue; this
> removes at least one desync cause.
Thanks! It's good that this issue is so alive latelly :) If there's an
idea that some debugging environment should be built in python or such,
I could be part of it.

-- 
Karol Kreński

**Comment by timfel** [comment]
@Andrettin I had replaced all the MyRand with SyncRand calls in the network-issues branch, that doesn't fix it. Also, afaict unit direction isn't at all used in determining unit choices. The problem i'm tracing it back to right now is sub-tile movement, which I traced further back to floating-point randomness.

**Comment by timfel** [comment]
And (also on that branch) I fixed the RandomSeed changing, because that isn't actually deterministic either, even though it's just dealing with unsigned. The bit-width of that type is *not* fixed across compilers and platforms, and overflow behavior is *not* the same at different optimization levels of gcc.

**Comment by timfel** [comment]
If you dump all unit info from the server and clients (`-i`) on a map that is ~60% full of units that are moving about, I find sub-tile deviations already after just 200 cycles.

**Comment by timfel** [comment]
Also, desyncs happen even without critters, in an otherwise empty map with only peasants and trees. The desync happens when the sub-tile differences amount to enough so that the AI sends a different peasant to harvest.

**Comment by timfel** [comment]
also, I have not been able to reproduce this when I run *only* Visual Studio debug builds. One reason may be that floating point arithmetic is by default forced to be precise, even if that is slower. On gcc, on the other hand, it is not so easy to force that.

**Comment by mimooh** [comment]
> Also, desyncs happen even without critters, in an otherwise empty map
No critters, but the tiles this time, fine. But we haven't looked at
critters potential just yet :) Basically we should be expecting multiple
causes to out-of-sync - this is what other developers experienced.

-- 
Karol Kreński

**Comment by Andrettin** [comment]
@timfel Great to hear that you've made so much progress in pinpointing the issue!

One question about this, though:
> And (also on that branch) I fixed the RandomSeed changing

Wouldn't that mean that SyncRand() will always return the same number?

**Comment by timfel** [comment]
@Andrettin no, I merely changed SyncRand to use 32bit clamped operations (without the multiplication, which could previously result in undefined behavior due to overflowing and then clamping on the network)

**Comment by timfel** [comment]
See this commit: https://github.com/Wargus/stratagus/commit/b1b89544da3bfa49ba754f2aa0bd3ee3b3079ceb

**Comment by Andrettin** [comment]
Ah, I see, thanks :)

**Comment by mimooh** [comment]
> Wouldn't that mean that SyncRand() will always return the same number?
I don't know if it comes from SyncRand(), but there is some fixed
number. This is my log_of_stratagus_0.log in stratagus (network-issues
branch): It starts from another number, but quickly stalls at 16700 in
two games I tried.

2: 17 unit-peasant 1 P0 Refs 1: 16700 56,110 0,0
2: 18 unit-peasant 1 P0 Refs 1: 16700 57,110 0,0
2: 19 unit-peasant 1 P0 Refs 1: 16700 56,111 0,0
2: 20 unit-peasant 1 P0 Refs 1: 16700 57,111 0,0
(...)
14529: 39 unit-critter 1 P15 Refs 1: 16700 30,66 0,0
14529: 40 unit-oil-patch 1 P15 Refs 1: 16700 89,49 0,0
14529: 41 unit-oil-patch 1 P15 Refs 1: 16700 17,113 0,0

I move my units around, but 16700 never changes. And it wasn't like that
before - 16700 was changing to various values.

> @timfel Great to hear that you've made so much progress in pinpointing the issue!
That's right Tim! :)

-- 
Karol Kreński

**Comment by timfel** [comment]
@mimooh syncrand still needs tweaking, that stalling is expected with the current code, i just wanted to see if the general approach would change anything.

@Andrettin I have printf-debugged my way back from the desync I'm seeing to AStar. The diff below compares log from client and server. You can see AStar generating different paths, which eventually leads to the unit (number 738) moving differently and ending up on different tiles, causing the desync:
```diff
 20: 738 unit-peasant Getting resource at P12 to -1,-1 or 194,68
-20: 738 unit-peasant P12: AStarFindPath pos:204,66 goal:193,67 goalsz:3,3 unitsz:1,1, range:0,1, path: { 6, 6, 7, 5, 6, 6, 6, 5, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 }
-20: 738 unit-peasant Moving P12, output Path is: { 6, 6, 7, 5, 6, 6, 6, 5 }
+20: 738 unit-peasant P12: AStarFindPath pos:204,66 goal:193,67 goalsz:3,3 unitsz:1,1, range:0,1, path: { 5, 6, 7, 5, 6, 6, 6, 5, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 }
+20: 738 unit-peasant Moving P12, output Path is: { 5, 6, 7, 5, 6, 6, 6, 5 }
 20: 738 unit-peasant Moving P12 through NextPathElement, result is: 8, posd.x/y: -1,1
 20: 738 unit-peasant Moving P12 posdx/y: -1,1 before heading up ix/y: 32,-32
 20: 738 unit-peasant Moving P12 posdx/y: -1,1 after heading up ix/y: 32,-32
 69: 738 unit-peasant Getting resource at P12 to -1,-1 or 194,68
-69: 738 unit-peasant Moving P12, output Path is: { 6, 6, 7, 5, 6, 6, 6 }
+69: 738 unit-peasant Moving P12, output Path is: { 5, 6, 7, 5, 6, 6, 6 }
 69: 738 unit-peasant Moving P12 through NextPathElement, result is: 7, posd.x/y: -1,0
 69: 738 unit-peasant Moving P12 posdx/y: -1,0 before heading up ix/y: 32,0
 69: 738 unit-peasant Moving P12 posdx/y: -1,0 after heading up ix/y: 32,0
 118: 738 unit-peasant Getting resource at P12 to -1,-1 or 194,68
-118: 738 unit-peasant Moving P12, output Path is: { 6, 6, 7, 5, 6, 6 }
+118: 738 unit-peasant Moving P12, output Path is: { 5, 6, 7, 5, 6, 6 }
 118: 738 unit-peasant Moving P12 through NextPathElement, result is: 6, posd.x/y: -1,0
 118: 738 unit-peasant Moving P12 posdx/y: -1,0 before heading up ix/y: 32,0
 118: 738 unit-peasant Moving P12 posdx/y: -1,0 after heading up ix/y: 32,0
 167: 738 unit-peasant Getting resource at P12 to -1,-1 or 194,68
-167: 738 unit-peasant Moving P12, output Path is: { 6, 6, 7, 5, 6 }
+167: 738 unit-peasant Moving P12, output Path is: { 5, 6, 7, 5, 6 }
 167: 738 unit-peasant Moving P12 through NextPathElement, result is: 5, posd.x/y: -1,0
 167: 738 unit-peasant Moving P12 posdx/y: -1,0 before heading up ix/y: 32,0
 167: 738 unit-peasant Moving P12 posdx/y: -1,0 after heading up ix/y: 32,0
 216: 738 unit-peasant Getting resource at P12 to -1,-1 or 194,68
-216: 738 unit-peasant Moving P12, output Path is: { 6, 6, 7, 5 }
+216: 738 unit-peasant Moving P12, output Path is: { 5, 6, 7, 5 }
 216: 738 unit-peasant Moving P12 through NextPathElement, result is: 4, posd.x/y: -1,1
 216: 738 unit-peasant Moving P12 posdx/y: -1,1 before heading up ix/y: 32,-32
 216: 738 unit-peasant Moving P12 posdx/y: -1,1 after heading up ix/y: 32,-32
 265: 738 unit-peasant Getting resource at P12 to -1,-1 or 194,68
-265: 738 unit-peasant Moving P12, output Path is: { 6, 6, 7 }
+265: 738 unit-peasant Moving P12, output Path is: { 5, 6, 7 }
 265: 738 unit-peasant Moving P12 through NextPathElement, result is: 3, posd.x/y: -1,-1
 265: 738 unit-peasant Moving P12 posdx/y: -1,-1 before heading up ix/y: 32,32
 265: 738 unit-peasant Moving P12 posdx/y: -1,-1 after heading up ix/y: 32,32
 314: 738 unit-peasant Getting resource at P12 to -1,-1 or 194,68
-314: 738 unit-peasant Moving P12, output Path is: { 6, 6 }
+314: 738 unit-peasant Moving P12, output Path is: { 5, 6 }
 314: 738 unit-peasant Moving P12 through NextPathElement, result is: 2, posd.x/y: -1,0
 314: 738 unit-peasant Moving P12 posdx/y: -1,0 before heading up ix/y: 32,0
 314: 738 unit-peasant Moving P12 posdx/y: -1,0 after heading up ix/y: 32,0
 363: 738 unit-peasant Getting resource at P12 to -1,-1 or 194,68
-363: 738 unit-peasant Moving P12, output Path is: { 6 }
-363: 738 unit-peasant Moving P12 through NextPathElement, result is: 1, posd.x/y: -1,0
-363: 738 unit-peasant Moving P12 posdx/y: -1,0 before heading up ix/y: 32,0
-363: 738 unit-peasant Moving P12 posdx/y: -1,0 after heading up ix/y: 32,0
-363: 738 unit-peasant Moving P12 posdx/y: -1,0 move:192 unitDir:4, ix/y: 32,0
-363: 738 unit-peasant 19 P12 Refs 3: 16700 196,67 28,0
+363: 738 unit-peasant Moving P12, output Path is: { 5 }
+363: 738 unit-peasant Moving P12 through NextPathElement, result is: 1, posd.x/y: -1,1
+363: 738 unit-peasant Moving P12 posdx/y: -1,1 before heading up ix/y: 32,-32
+363: 738 unit-peasant Moving P12 posdx/y: -1,1 after heading up ix/y: 32,-32
+363: 738 unit-peasant Moving P12 posdx/y: -1,1 move:160 unitDir:4, ix/y: 32,-32
+363: 738 unit-peasant 19 P12 Refs 3: 16700 196,68 28,-28
```

**Comment by Andrettin** [comment]
@timfel Ah, I see. Is there any reason we can't use integers instead of floats in those calculations?

**Comment by timfel** [comment]
I don't see where we do use floats in there, a quick search didn't show anything obvious to me.

**Comment by mimooh** [comment]
Is sin() not problematic for stratagus?

https://randomascii.wordpress.com/2013/07/16/floating-point-determinism/
In 32-bit Ubuntu 12.04 with glibc 2.15 the run-time sin() would use fsin making
for significant differences depending on whether sin() was calculated at
run-time or compile-time. On Ubuntu 12.04 this code does not print 1.0:

    const double pi_d = 3.14159265358979323846;
    const int zero = argc / 99;
    printf(“%f\n”, sin(pi_d + zero) / sin(pi_d));

    0.999967

Stratagus chunkparticle.cpp uses:
	float radians = deg2rad(MyRand() % 360);
	direction.y = sin(radians);

**Comment by mimooh** [comment]
Also, I guess stratagus is not having this, seems very useful feature:

http://stackoverflow.com/questions/20963419/cross-platform-floating-point-consistancy
In one of my own projects, during development, I used to hash all the relevant state (including a lot of floating-point numbers) in all the instances of the game and send the hash across the network each frame to make sure even one bit of that state wasn't different on different machines. This also helped with debugging, where instead of trusting my eyes to see when and where inconsistencies existed (which wouldn't tell me where they originated, anyways) I would know the instant some part of the state of the game on one machine started diverging from the others, and know exactly what it was (if the hash check failed, I would stop the simulation and start comparing the whole state.)

Does stratagus have a log of what is sent to the net? The -i -p are about logging complete game state and this is different from what goes to the network, right?

These beautiful logs of timfel yesterday differ from my logs. I guess they are not from -i -p?

**Comment by Andrettin** [comment]
@mimooh

I don't think chunkparticle.cpp and the other particle code is affecting the network - AFAIK what they do is purely graphical. And Wargus doesn't (last I saw) even make use of the particles, that's something out of the Stratagus games only Doom Wars does.

What you mentioned of hashing the entire relevant state sounds useful, though I'm not sure where to start if we were to do that to Stratagus.

@timfel Ah, I see, I thought there were floats in the movement calculation. By the way, there is a bug in Stratagus which affects unit facing (it doesn't round the unit facing properly), which I fixed for Wyrmsun. Do you think it could be related to the desync issue?

**Comment by Andrettin** [comment]
Another thing: I just noticed that for Wyrmsun I added the following code in CreateGame (in game.cpp), below "InitPlayers();":

```
	if (IsNetworkGame()) { // if is a network game, it is necessary to reinitialize the syncrand variables before beginning to load the map, due to random map generation
		SyncHash = 0;
		InitSyncRand();
	}
```

**Comment by Andrettin** [comment]
Regarding what I mentioned previously about the bug in facing with Stratagus - the direction is sometimes set to a value that is not a multiple of 32 (256 / 8 = 32, since there are 8 directions). This causes some weirdness in the game. I fixed that by adding the following code at the beginning of UnitUpdateHeading (in unit.cpp):

```
	//fix direction if it does not correspond to one of the defined directions
	int num_dir = std::max<int>(8, unit.Type->NumDirections);
	if (unit.Direction % (256 / num_dir) != 0) {
		unit.Direction = unit.Direction - (unit.Direction % (256 / num_dir));
	}
```

In UnitHeadingFromDeltaXY (in unit.cpp as well), I also replaced "unit.Direction = DirectionToHeading(delta);" with the following code (but it looks superfluous, I probably added this here before implementing the previous fix):

```
	int num_dir = std::max<int>(8, unit.Type->NumDirections);
	int heading = DirectionToHeading(delta) + ((256 / num_dir) / 2);
	if (heading % (256 / num_dir) != 0) {
		heading = heading - (heading % (256 / num_dir));
	}
	unit.Direction = heading;
```

**Comment by timfel** [comment]
@Andrettin sounds like these fixes should be in Stratagus, too. I can check if they help when I get home in ~10hs

**Comment by timfel** [comment]
Nope, that didn't fix it

**Comment by mimooh** [comment]
When it comes to AI harvesting perhaps some inspirations can be found here:
http://www.codeofhonor.com/blog/the-starcraft-path-finding-hack

Also, if it's not floats in stratagus then the other suspicious thing is uninitialized variables:
http://www.gamasutra.com/view/news/126022/Opinion_Synchronous_RTS_Engines_And_A_Tale_of_Desyncs.php

**Comment by timfel** [comment]
Ok, should've used valgrind more thoroughly, rather than just static analysis tools.
```valgrind
==8964== Source and destination overlap in memcpy(0x1953c098, 0x1953c0a0, 232)
==8964==    at 0x4C2F71C: memcpy@@GLIBC_2.14 (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==8964==    by 0x5EBE1C: AStarReplaceNode(int) (astar.cpp:467)
==8964==    by 0x5ECE8D: AStarFindPath(Vec2T<short> const&, Vec2T<short> const&, int, int, int, int, int, int, char*, int, CUnit const&) (astar.cpp:1068)
==8964==    by 0x5EE8D7: NewPath(PathFinderInput&, PathFinderOutput&) (pathfinder.cpp:340)
==8964==    by 0x5EEBF7: NextPathElement(CUnit&, short*, short*) (pathfinder.cpp:395)
==8964==    by 0x527BAD: DoActionMove(CUnit&) (action_move.cpp:150)
==8964==    by 0x52BAE4: COrder_Resource::MoveToResource_Terrain(CUnit&) (action_resource.cpp:375)
==8964==    by 0x52BD99: COrder_Resource::MoveToResource(CUnit&) (action_resource.cpp:454)
==8964==    by 0x52E67B: COrder_Resource::Execute(CUnit&) (action_resource.cpp:1263)
==8964==    by 0x5370BD: HandleUnitAction(CUnit&) (actions.cpp:401)
==8964==    by 0x53774A: void UnitActionsEachCycle<__gnu_cxx::__normal_iterator<CUnit**, std::vector<CUnit*, std::allocator<CUnit*> > > >(__gnu_cxx::__normal_iterator<CUnit**, std::vector<CUnit*, std::allocator<CUnit*> > >, __gnu_cxx::__normal_iterator<CUnit**, std::vector<CUnit*, std::allocator<CUnit*> > >) (actions.cpp:492)
==8964==    by 0x53743F: UnitActions() (actions.cpp:539)
```

After fixing this, I no longer get desyncs on my test map. The change is simply this:
```diff
-       //memmove(&OpenSet[pos], &OpenSet[pos+1], sizeof(Open) * (OpenSetSize-pos));
-       memcpy(&OpenSet[pos], &OpenSet[pos + 1], sizeof(Open) * (OpenSetSize - pos));
+       memmove(&OpenSet[pos], &OpenSet[pos+1], sizeof(Open) * (OpenSetSize-pos));
```

The previous commit that commented out memmove and replaced it with memcpy was in 2008: 5b6bc1aee657fe3e6468a16fa6fc72022194b7b0. Before that, the memmove has been in there since 2006: db93e5551df933631d0f25f9e8967cc751671748

**Comment by timfel** [comment]
So, @mimooh, I guess I'd be glad for you to try the branch again and see if it still desyncs for you.

**Comment by Andrettin** [comment]
@timfel that's great, thank you very much! :)

**Comment by Andrettin** [comment]
It turns out this fix was already in Wyrmgus, which is why I wasn't experiencing the bug with Wyrmsun. For reference, here's the commit:
https://github.com/Andrettin/Wyrmgus/commit/ddd1db42e913cc7df68828469680d5b153132dbe#diff-8fbd44e1a5fa0d06d4e7834c3fb46d60

**Comment by mimooh** [comment]
The two of you are just wizards! :) Thank you very much and I hope to
have it tested later tonight.

-- 
Karol Kreński

**Comment by mimooh** [comment]
Not a single desync on a heavy localhost game!  There were lot's of desyncs previously. Will try a windows vs linux game against a friend, but it looks very promising :) Thanks again for your great work!

**Comment by mimooh** [comment]
Can we have the latest network fixes merged into upstream? I'd like to have it built for windows for the testing.

**Comment by mizukami999** [comment]
Simply amazing!

**Comment by DinkyDyeAussie** [comment]
Is it working on windows?

**Comment by mimooh** [comment]
@DinkyDyeAussie: yes, the changes are merged to the official version, so you can try the exe at the bottom of this page:
https://github.com/Wargus/stratagus

Let us know if you have any desyncs. I will have it tested in a windows against linux game this week hopefully.

**Comment by a-detiste** [comment]
will test on linux armv7 versus linux amd64

**Comment by mimooh** [comment]
Can I run a remote server (over ssh perhaps) and then have me and a friend connect to it? There's a 'dedicated' option, which sounds like it, but it talks about some AI clients. I'd like to play a human vs human game.

**Comment by timfel** [comment]
@mimooh It's not really dedicated, it will run as an AI player. There is no human vs human dedicated server option.

**Comment by Andrettin** [comment]
@timfel Speaking of servers, do you know what a metaserver is, or how such a thing works? Stratagus has unfinished metaserver code, which could be further developed later on.

**Comment by timfel** [comment]
@Andrettin I have no idea how the metaserver was supposed to work, I haven't looked at that code

**Comment by Andrettin** [comment]
@timfel But do you know what a "metaserver" even is? I have no idea, and I couldn't find much information on it on the internet either :(

**Comment by mimooh** [comment]
It is a server for listing games and allowing people to play together as I understand it. network.cpp relates metaserver to bnetd, which is a battle.net-like software.

**Comment by mimooh** [comment]
The game works just perfect in network. Today we had two 4 players games under ubuntu, which were not disturbed by out-of-sync problems, not even once. So I'd say the issue is pretty much fixed. Thanks once more for sorting this issue out! :)

**Comment by elicoten** [comment]
Firstly thanks so much to @timfel @Andrettin  and @mimooh for getting this sorted. Really really appreciated by me :video_game:  

You say that the download link is at the bottom of the main page of the repository. When I try that takes it takes me to https://github.com/Wargus/stratagus/releases which says that the "master-build" release hasn't been modified since 30 November, so presumably this can't be it - can it?

Also what about the Ubuntu release - on https://launchpad.net/~stratagus/+archive/ubuntu/ppa the latest version shown is 2.4.0-0~1968~ubuntu15.10.1 which is dated 8th January - so this too surely can't be the right one?

Thanks again

**Comment by mimooh** [comment]
@elicoten, as it comes to ubuntu, the change was in stratagus only - wargus package was fine all the time. As it comes to windows, I guess wargus displays wrong dates, because I installed wargus from there and the bug which I reported in January 2016 was gone.

**Comment by timfel** [comment]
I've addressed three more of the above that were specific to Wargus. The preparation room and map quality thing I'm ignoring for now. Adding a few well-made high-quality maps could be a good task for someone who likes making maps :)
